tests/ut/test_feder.cc

```diff
@@ -0,0 +1,234 @@
+                REQUIRE(n < nb);
+            }
+        }
+    }
+
+    // check IDSet
+    std::unordered_set<int64_t> id_set;
+    knowhere::Json j2 = nlohmann::json::parse(json_id_set);
+    nlohmann::from_json(j2, id_set);
+    std::cout << "id_set num = " << id_set.size() << std::endl;
+    for (auto id : id_set) {
+        REQUIRE(id >= 0);
+        REQUIRE(id < nb);
+    }
+}
+
+void
+CheckHnswVisitInfo(const knowhere::DataSetPtr result, int64_t nb) {
+    auto json_info = result->GetJsonInfo();
+    auto json_id_set = result->GetJsonIdSet();
+    // std::cout << json_info << std::endl;
+    std::cout << "json_info size = " << json_info.size() << std::endl;
+    std::cout << "json_id_set size = " << json_id_set.size() << std::endl;
+
+    // check HNSWVisitInfo
+    knowhere::feder::hnsw::HNSWVisitInfo visit_info;
+    knowhere::Json j1 = nlohmann::json::parse(json_info);
+    nlohmann::from_json(j1, visit_info);
+
+    for (auto& level_visit_record : visit_info.GetInfos()) {
+        auto& records = level_visit_record.GetRecords();
+        for (auto& record : records) {
+            auto id_from = std::get<0>(record);
+            auto id_to = std::get<1>(record);
+            auto dist = std::get<2>(record);
+            REQUIRE(id_from >= 0);
+            REQUIRE(id_to >= 0);
+            REQUIRE(id_from < nb);
+            REQUIRE(id_to < nb);
+            REQUIRE((dist >= 0.0 || dist == -1.0));
+        }
+    }
+
+    // check IDSet
+    std::unordered_set<int64_t> id_set;
+    knowhere::Json j2 = nlohmann::json::parse(json_id_set);
+    nlohmann::from_json(j2, id_set);
+    std::cout << "id_set num = " << id_set.size() << std::endl;
+    for (auto id : id_set) {
+        REQUIRE(id >= 0);
+        REQUIRE(id < nb);
+    }
+}
+
+void
+CheckIvfFlatMeta(const knowhere::DataSetPtr result, int64_t nb, const knowhere::Json& cfg) {
+    auto json_info = result->GetJsonInfo();
+    auto json_id_set = result->GetJsonIdSet();
+    // std::cout << json_info << std::endl;
+    std::cout << "json_info size = " << json_info.size() << std::endl;
+    std::cout << "json_id_set size = " << json_id_set.size() << std::endl;
+
+    // check IVFFlatMeta
+    knowhere::feder::ivfflat::IVFFlatMeta meta;
+    knowhere::Json j1 = nlohmann::json::parse(json_info);
+    nlohmann::from_json(j1, meta);
+
+    REQUIRE(meta.GetNlist() == cfg[knowhere::indexparam::NLIST]);
+    REQUIRE(meta.GetDim() == cfg[knowhere::meta::DIM]);
+    REQUIRE(meta.GetNtotal() == nb);
+
+    // sum of all cluster nodes should be equal to nb
+    auto& clusters = meta.GetClusters();
+    std::unordered_set<int64_t> all_id_set;
+    REQUIRE(clusters.size() == cfg[knowhere::indexparam::NLIST]);
+    for (auto& cluster : clusters) {
+        for (auto id : cluster.node_ids_) {
+            REQUIRE(id >= 0);
+            REQUIRE(id < nb);
+            all_id_set.insert(id);
+        }
+    }
+    REQUIRE(all_id_set.size() == (size_t)nb);
+
+    // check IDSet
```

Comment:
What do we check here? Are we only expecting the json operations to succeed? pls comment if so.

Reply:
update